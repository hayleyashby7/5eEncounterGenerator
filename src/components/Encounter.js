import React from 'react';

const Encounter = () => {



	return (
		<div>
			<h2>Encounter</h2>
		</div>
	);
};

export default Encounter;
