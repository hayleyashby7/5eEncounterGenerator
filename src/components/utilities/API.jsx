export const getMonstersFromAPI = async (challengeRating) => {
	return fetch(`https://api.open5e.com/monsters/?challenge_rating=${challengeRating}`).then((response) => {
		if (response.status === 200) return response.json();
		else throw new Error('Error fetching data');
	});
};
