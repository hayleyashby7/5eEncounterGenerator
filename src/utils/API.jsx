export const getMonstersFromAPI = async ({ challengeRating }) => {
    try {
        const response = await fetch(
            `https://api.open5e.com/monsters/?challenge_rating=${challengeRating}`,
        );
        return await response.json();
    } catch (e) {
        throw new Error(e);
    }
};

export default getMonstersFromAPI;
