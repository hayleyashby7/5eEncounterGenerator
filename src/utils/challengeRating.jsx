/* eslint-disable no-console */
/* eslint-disable no-fallthrough */

import xpThresholdssByCharLevel from './data/xpThresholdsByCharLevel.json';
import xpByChallengeRating from './data/xpByChallengeRating.json';

const xpMultiplier = (xp, numCharacters) => {
    let totalXP = xp;

    if (numCharacters < 3) {
        totalXP *= 1.5;
    }

    if (numCharacters > 5) {
        totalXP *= 0.5;
    }

    return totalXP;
};

const getClosestChallengeRating = (data, targetRange, numCharacters) => {
    let start = 0;
    let end = data.length - 1;
    let middle;
    let testValue;
    let closestCR = null;

    while (start < end && closestCR === null) {
        // Update middle value
        middle = Math.floor((start + end) / 2);
        testValue = xpMultiplier(data[middle].XP, numCharacters);

        switch (true) {
            case testValue > targetRange[1]:
                // Value is above range - continue search
                end = middle + 1;
                break;

            case testValue < targetRange[0]:
                // Value is below range - continue search
                start = middle - 1;
                break;

            case testValue === targetRange[0] || testValue === targetRange[1] - 1:
            // Exact match - fall through
            case testValue > targetRange[0] && testValue < targetRange[1]:
                // Value is within range
                closestCR = data[middle].CR;
                break;

            default:
                // Error if reached here
                console.error('Error: Cannot calculate Challenge Rating - defaulting to CR 0');
                closestCR = 0;
                break;
        }
    }

    if (closestCR === null) {
        // No exact match, return the nearest value
        if (testValue > targetRange[1]) {
            closestCR = data[middle - 1].CR;
        } else {
            closestCR = data[middle].CR;
        }
    }

    return closestCR;
};

export const isValidChallengeRating = ({ challengeRating }) => {
    if (challengeRating >= 0 && challengeRating <= 30) {
        return true;
    }
    return false;
};

export const calculateChallengeRating = (numCharacters, level, difficulty) => {
    if (level > 20 || level < 1) {
        console.error(
            `Error: Cannot calculate Challenge Rating - Invalid character level: ${level}`,
        );
        return null;
    }

    if (numCharacters < 1 || numCharacters > 10) {
        console.error(
            `Error: Cannot calculate Challenge Rating - Invalid number of characters: ${numCharacters}`,
        );
        return null;
    }

    // Determine XP Threshold for character level by difficulty
    const diff = xpThresholdssByCharLevel.level[level - 1].difficulty;

    // Get XP Thresholds [min, max] for difficulty
    let xpThresholds;

    switch (difficulty) {
        case 'Easy':
            xpThresholds = [diff.easy, diff.medium];
            break;
        case 'Medium':
            xpThresholds = [diff.medium, diff.hard];
            break;
        case 'Hard':
            xpThresholds = [diff.hard, diff.deadly];
            break;
        case 'Deadly':
            xpThresholds = [diff.deadly, diff.max];
            break;
        default:
            xpThresholds = null;
    }

    if (xpThresholds === null) {
        console.error(
            `Error: Cannot calculate Challenge Rating - Invalid difficulty ${difficulty}`,
        );
        return null;
    }

    // Determine the total party thresholds
    const partyXPThresholds = [numCharacters * xpThresholds[0], numCharacters * xpThresholds[1]];

    // Convert JSON to array keeping original order
    const xpByChallengeRatingArray = Object.keys(xpByChallengeRating).map(
        (key) => xpByChallengeRating[key],
    );

    // Return the closest Challenge Rating based on range
    return getClosestChallengeRating(xpByChallengeRatingArray, partyXPThresholds, numCharacters);
};
