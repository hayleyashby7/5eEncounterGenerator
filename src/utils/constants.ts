const { VITE_DDEG_API_KEY: API_KEY } = import.meta.env;

export { API_KEY };
