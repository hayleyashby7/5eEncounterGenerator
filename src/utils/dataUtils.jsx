export const removeUnderscores = (str) => {
    return str.replace(/_/g, ' ');
};

export const capitalise = (str) => {
    return str.charAt(0).toUpperCase() + str.slice(1).toLowerCase();
};

export const formatString = (str) => {
    return capitalise(removeUnderscores(str));
};
