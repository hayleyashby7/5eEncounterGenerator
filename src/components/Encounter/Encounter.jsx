import React, { useEffect, useState } from 'react';
import PropTypes from 'prop-types';
import { getMonstersFromAPI } from '../../utils/API';
import isValidChallengeRating from '../../utils/challengeRating';

function Encounter({ challengeRating }) {
    const [monsters, setMonsters] = useState([]);
    const [error, setError] = useState(null);
    const [message, setMessage] = useState('Loading...');

    useEffect(() => {
        const fetchData = async () => {
            try {

                if (!isValidChallengeRating({ challengeRating })) {
                    throw new Error('Unable to request data.');
                }
                
                const data = await getMonstersFromAPI({ challengeRating });

                if (data instanceof Error) {
                    throw new Error(
                        'Unable to retrieve monsters from server. Please try again later.',
                    );
                }
                if (data.count === 0) {
                    throw new Error('No suitable monsters found.');
                }

                setMonsters(data.results);
            } catch (err) {
                setError(true);
                setMessage(err.message);
            }
        };

        fetchData();
    }, [challengeRating]);

    return (
        <>
            <h1>List of Monsters</h1>
            {error && <div>{message}</div>}{' '}
            {monsters ? (
                <ul>
                    {' '}
                    {monsters.map((monster) => (
                        <li key={monster.slug}>{monster.name}</li>
                    ))}{' '}
                </ul>
            ) : (
                <p>{message}</p>
            )}{' '}
        </>
    );
}

Encounter.propTypes = { challengeRating: PropTypes.number };
Encounter.defaultProps = { challengeRating: null };

export default Encounter;
