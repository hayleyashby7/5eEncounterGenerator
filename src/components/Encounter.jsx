import React, { useEffect, useState } from 'react';
import PropTypes from 'prop-types';
import Row from './Table/Row';
import { getMonstersFromAPI } from '../utils/API';
import { isValidChallengeRating } from '../utils/challengeRating';

function Encounter({ challengeRating }) {
    const [monsters, setMonsters] = useState([]);
    const [error, setError] = useState(null);
    const [message, setMessage] = useState('Loading...');

    useEffect(() => {
        const fetchData = async () => {
            try {
                if (!isValidChallengeRating({ challengeRating })) {
                    throw new Error('Unable to request data.');
                }

                const data = await getMonstersFromAPI({ challengeRating });

                if (data instanceof Error) {
                    throw new Error(
                        'Unable to retrieve monsters from server. Please try again later.',
                    );
                }
                if (data.count === 0) {
                    throw new Error('No suitable monsters found.');
                }

                setMonsters(data.results);
                setError(false);
            } catch (err) {
                setError(true);
                setMonsters([]);
                setMessage(err.message);
            }
        };

        fetchData();
    }, [challengeRating]);

    return (
        <div className='bg-neutral-900 text-orange-100'>
            <h1>List of Monsters</h1>
            {error && <div>{message}</div>}{' '}
            {monsters ? (
                <table>
                    <thead>
                        <tr>
                            <th>Name</th>
                            <th>CR</th>
                            <th>Type</th>
                            <th>Size</th>
                            <th>Alignment</th>
                        </tr>
                    </thead>
                    <tbody>
                        {monsters.map((monster) => (
                            <Row key={monster.slug} data={monster} />
                        ))}
                    </tbody>
                </table>
            ) : (
                <p>{message}</p>
            )}{' '}
        </div>
    );
}

Encounter.propTypes = { challengeRating: PropTypes.number };
Encounter.defaultProps = { challengeRating: -1 };

export default Encounter;
