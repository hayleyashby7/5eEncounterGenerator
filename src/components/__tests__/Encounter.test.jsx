import { vi, it, describe, expect, afterEach } from 'vitest';
import React from 'react';
import { render, screen } from '@testing-library/react';
import Encounter from '../Encounter';
import { getMonstersFromAPI } from '../../utils/API';

vi.mock('../../utils/API');

describe('Encounter', () => {
    afterEach(() => {
        vi.clearAllMocks();
    });

    it('should render a list of monsters with successful API call', async () => {
        // Arrange

        // Mock the API call
        getMonstersFromAPI.mockReturnValue(
            Promise.resolve({
                count: 2,
                results: [
                    {
                        slug: 'goblin',
                        name: 'Goblin',
                        challenge_rating: '1',
                        type: 'Humanoid',
                        size: 'Small',
                        alignment: 'Neutral',
                    },
                    {
                        slug: 'orc',
                        name: 'Orc',
                        challenge_rating: '1',
                        type: 'Humanoid',
                        size: 'Medium',
                        alignment: 'Chaotic',
                    },
                ],
            }),
        );

        // Act
        render(<Encounter challengeRating={1} />);

        // Assert
        // Heading is displayed
        expect(screen.getByRole('heading')).toHaveTextContent(/List of Monsters/i);

        // Mocked data is displayed
        expect(await screen.findByText(/Goblin/i)).toBeInTheDocument();
        expect(await screen.findByText(/Orc/i)).toBeInTheDocument();

        // No error message is displayed
        expect(screen.queryByText(/No monsters found/i)).not.toBeInTheDocument();
    });

    it('should show an error message with a failed API call', async () => {
        // Arrange

        // Mock the API call
        getMonstersFromAPI.mockReturnValue(Promise.reject(new Error('API call failed')));

        // Act
        render(<Encounter challengeRating={1} />);

        // Assert
        // Heading is still displayed
        expect(screen.getByRole('heading')).toHaveTextContent(/List of Monsters/i);

        // Error message is displayed
        expect(await screen.findByText(/API call failed/i)).toBeInTheDocument();

        // Mocked data is not displayed
        expect(screen.queryByText(/Goblin/i)).not.toBeInTheDocument();
        expect(screen.queryByText(/Orc/i)).not.toBeInTheDocument();
    });

    it('should render a message to advise if API call resolves but an error is returned', async () => {
        // Arrange

        // Mock the API call
        getMonstersFromAPI.mockReturnValue(Promise.resolve(new Error('API call failed')));

        // Act
        render(<Encounter challengeRating={1} />);

        // Assert
        // Heading is displayed
        expect(screen.getByRole('heading')).toHaveTextContent(/List of Monsters/i);

        // Error message is displayed
        expect(
            await screen.findByText(
                /Unable to retrieve monsters from server. Please try again later./i,
            ),
        ).toBeInTheDocument();

        // Mocked data is not displayed
        expect(screen.queryByText(/Goblin/i)).not.toBeInTheDocument();
        expect(screen.queryByText(/Orc/i)).not.toBeInTheDocument();
    });

    it('should render a message to advise if no data recieved on a successful API call', async () => {
        // Arrange

        // Mock the API call
        getMonstersFromAPI.mockReturnValue(
            Promise.resolve({
                count: 0,
                results: [],
            }),
        );

        // Act
        render(<Encounter challengeRating={1} />);

        // Assert
        // Heading is displayed
        expect(screen.getByRole('heading')).toHaveTextContent(/List of Monsters/i);

        // Error message is displayed
        expect(await screen.findByText(/No suitable monsters found./i)).toBeInTheDocument();

        // Mocked data is not displayed
        expect(screen.queryByText(/Goblin/i)).not.toBeInTheDocument();
        expect(screen.queryByText(/Orc/i)).not.toBeInTheDocument();
    });

    it('should not make API call if no challenge rating passed to component', async () => {
        // Arrange

        // Mock the API call
        getMonstersFromAPI.mockReturnValue(
            Promise.resolve({
                count: 2,
                results: [
                    {
                        slug: 'goblin',
                        name: 'Goblin',
                        challenge_rating: '1',
                        type: 'Humanoid',
                        size: 'Small',
                        alignment: 'Neutral',
                    },
                    {
                        slug: 'orc',
                        name: 'Orc',
                        challenge_rating: '1',
                        type: 'Humanoid',
                        size: 'Medium',
                        alignment: 'Chaotic',
                    },
                ],
            }),
        );

        // Act
        render(<Encounter />);

        // Assert
        // Heading is displayed
        expect(screen.getByRole('heading')).toHaveTextContent(/List of Monsters/i);

        // Error message is displayed
        expect(await screen.findByText(/Unable to request data./i)).toBeInTheDocument();

        // Mocked data is not displayed
        expect(screen.queryByText(/Goblin/i)).not.toBeInTheDocument();
        expect(screen.queryByText(/Orc/i)).not.toBeInTheDocument();
    });
});
