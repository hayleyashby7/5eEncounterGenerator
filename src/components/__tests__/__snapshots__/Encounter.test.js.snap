// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Encounter has correct header and content renders Encounter heading 1`] = `
<div>
  <div>
    <h2>
      Encounter
    </h2>
  </div>
</div>
`;
